### ISSUE TITLE
Provide compatibility with OpenJ9

### ISSUE THREAD

**Issue by mcs6502** [OpenHFT/Chronicle-Queue#599]
Please could you provide the ability to run Chronicle Queue on OpenJ9 VM.

**Comment by mcs6502** [comment]
Chronicle Queue version 5.17.8 fails when running on OpenJ9, getting a `java.lang.IllegalStateException: expecting index, was index2` exception in `SCQIndexing.array()`.
The underlying problem is that the `StringUtils.extractBytes()` call expects the String class to contain an array of single-byte characters, whereas on OpenJ9 it has a byte array with UTF-16 in it (uncompressed).
This particular issue could be raised in Chronicle-Core, but I felt providing OpenJ9 compatibility could have wider implications than simply fixing the StringUtils class.

**Comment by RobAustin** [comment]
Are you able to create a pull request for chronicle-core and apply the fix you suggest ?

**Comment by dpisklov** [comment]
@mcs6502 as a small company, we are unable to provide support for the exotic JVM versions our clients do not use. Please feel free to provide a pull request to add this support, or consider becoming our customer if you require this for commercial purposes. In latter case please contact sales@chronicle.software

**Comment by mcs6502** [comment]
Rob and Dmitry -- I have researched this a bit further, and found the following assertion in `shouldExtractBytesFromString()` method in StringUtilsTest:
```
"Is this test running on JDK9 with compact strings disabled?"
```
This indicates that the `byte[]` methods on StringUtils are designed to work only on those Java 9+ VMs that have the compact strings feature enabled, but not on ones that have non-compact strings. This is not specific to OpenJ9, and HotSpot should fail this test as well (but it doesn't because compact strings are enabled by default).
Conversely, OpenJ9 should be able to run Chronicle Queue with compact strings. Indeed I can confirm (after very limited testing) that Chronicle Queue is able to work on OpenJ9 VM with the `-XX:+CompactStrings` option enabled.

Now the `extractBytes()` method in StringUtils doesn't need any fixing -- it returns the underlying byte array as requested. However, the callers of this method (for example, `MappedBytes.write8bit()` via `append8bit0()`) should be modified to also query the string's `coder` and act accordingly -- for instance, by doubling the length for UTF-16. Actually, it is more involved than just doubling the length because the coder flag (0/1) should be persisted with string bytes so that the correct string can be reconstituted afterwards (for example, in the `newStringFromBytes()` method of StringUtils).

The above considerations are only relevant if Chronicle Queue needs to work on VMs with non-compact strings. Apparently, there is no such need at the moment -- otherwise you would have seen the same failure in `SCQIndexing` earlier. So for practical purposes, Chronicle Queue can be considered compatible with OpenJ9, provided the `-XX:+CompactStrings` option is used.

**Comment by dpisklov** [comment]
@mcs6502 fair enough - thanks for the info and research. Shall we close the issue then?

**Comment by RobAustin** [comment]
thank you we have updated our documentation to reflect your findings, thank you again.

see https://github.com/OpenHFT/Chronicle-Queue/blob/master/README.adoc#can-chronicle-queue-work-on-openj9 <https://github.com/OpenHFT/Chronicle-Queue/blob/master/README.adoc#can-chronicle-queue-work-on-openj9>

**Comment by mcs6502** [comment]
Might also be worth adding that OpenJ9 version 0.12.1 and earlier requires the file descriptor limit to be manually adjusted to a higher value -- for example, using the command `ulimit -Sn 500`.
And yes -- should be good to close this issue, thank you!

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.11](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.11), [BOM-2.17.206](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.206)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.11](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.11), [BOM-2.17.225](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.225)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.11](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.11), [BOM-2.17.241](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.241)
